python/revng/pypeline/storage/local_provider.py

```diff
@@ -504,3 +557,40 @@ def _compare_file(path: Path, request: FileRequest) -> bool:
             and (request.size is None or path.stat().st_size == request.size)
             and compute_hash(path) == request.hash
         )
+
+    def add_advanced_dependencies(
+        self, pipe_id: int, configuration_hash: str, data: PipeAdvancedInvalidation
+    ):
+        with self._cursor() as cursor:
+            for index, container_data in enumerate(data):
+                for object_id, invalidation_blob in container_data:
+                    cursor.execute(
+                        "REPLACE INTO advanced_dependencies VALUES (?, ?, ?, ?, ?)",
+                        (
+                            pipe_id,
+                            configuration_hash,
+                            index,
+                            object_id.to_bytes(),
+                            bytes(invalidation_blob),
+                        ),
+                    )
+            self._write_metadata(cursor)
+
+    def get_advanced_dependencies(
```

Comment:
This does not return dependencies but opaque data that enable you to compute dependencies.

`get_custom_invalidation_data`